pkg/metricsprovider/prometheus.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2020
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metricsprovider
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"github.com/paypal/load-watcher/pkg/watcher"
+	log "github.com/sirupsen/logrus"
+	"net/http"
+	"os"
+	"reflect"
+	"strconv"
+
+	// For out of cluster connections.
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
+)
+
+var (
+	promHost    string
+	promToken	string
+	promTokenPresent = false
+	node_metric_query = map[string]string{
+		watcher.CPU : 	"instance:node_cpu:ratio",
+		watcher.Memory : "instance:node_memory_utilization:ratio",
+	}
+
+)
+
+const (
+	// env variable that provides path to kube config file, if deploying from outside K8s cluster
+	promHostKey = "PROM_HOST"
+	promTokenKey = "PROM_TOKEN"
+	promQuery = "/api/v1/query?query="
+	prom_std_method = "stddev_over_time"
+	prom_avg_method = "avg_over_time"
+	prom_cpu_metric = "instance:node_cpu:ratio"
+	prom_mem_metric = "instance:node_memory_utilisation:ratio"
+)
+
+func init() {
+	var promHostPresent bool
+
+	promHost, promHostPresent = os.LookupEnv(promHostKey)
+	promToken, promTokenPresent = os.LookupEnv(promTokenKey)
+	if !promHostPresent {
+		promHost = "prometheus-k8s:9090"
+	}
+}
+
+type promClient struct {
+	client http.Client
```

Comment:
I would strongly recommend using native Prometheus [client](https://github.com/prometheus/client_golang), they seem to have pretty good support in Go lang. This will simplify lot of boilerplate code for json unmarshal etc. for decoding their payload as done below in functions. Also it will probably support connection pooling and other optimizations that will work better with Prometheus server.

Reply:
Makes sense. I will try the [Prometheus client](https://github.com/prometheus/client_golang/blob/master/api/prometheus/v1/api.go).